plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/cache/CachingHiveMetastore.java

```diff
@@ -1330,6 +1360,85 @@ private static <K, V> Cache<K, AtomicReference<V>> buildBulkCache(
         return cacheBuilder.build();
     }
 
+    private class TableNamesCache
+    {
+        private static final String DUMMY_CACHE_KEY = "$KEY";
+
+        @Nullable
+        private final LoadingCache<String, List<String>> perSchemaCache;
+        /**
+         * Normally, we would use a memoized supplier. However, there is an additional refresh
+         * capability so we use a single entry loading cache with a dummy key
+         */
+        @Nullable
+        private final LoadingCache<String, Map<String, List<String>>> batchCache;
+
+        private final boolean isBatchCache;
+
+        public TableNamesCache(
+                CacheFactory cacheFactory,
+                com.google.common.base.Function<String, List<String>> perSchemaLoader,
+                Supplier<Map<String, List<String>>> batchLoader)
+        {
+            isBatchCache = delegate.supportBatchListingOperations();
+
+            if (isBatchCache) {
+                batchCache = cacheFactory.buildCache(ignore -> batchLoader.get());
+                perSchemaCache = null;
+            }
+            else {
+                batchCache = null;
+                perSchemaCache = cacheFactory.buildCache(perSchemaLoader);
+            }
+        }
+
+        public List<String> get(String databaseName)
+        {
+            if (isBatchCache) {
+                // If we ask the delegated metastore for a non-existent schema it will return an empty list.
+                // However, if we store all existing schemas in a map, then null will be returned
+                return getAll().getOrDefault(databaseName, ImmutableList.of());
+            }
+            return CachingHiveMetastore.this.get(perSchemaCache, databaseName);
+        }
+
+        public Map<String, List<String>> getAll()
+        {
+            checkState(isBatchCache, "getAll method may only be invoked if the delegated metastore supports batchGetTables");
+            return CachingHiveMetastore.this.get(batchCache, DUMMY_CACHE_KEY);
+        }
+
+        public void invalidate(String databaseName)
+        {
+            if (isBatchCache) {
```

Comment:
do you need this if here, why not to invalidate all caches? Same below

Reply:
One of the caches is null, so I need to know which one to invalidate.